### PR TITLE
Fix static hooks

### PR DESCRIPTION
**Pull request by jordanbrauer**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| Fixed tickets | #351 

### Issue

The issue is essentially, exactly, as described by @fnzr in the related ticket (great job and good find!).

> Basically, the problem happens because each test, when it's being parsed, registers the global hook on a shared variable (Testable::$beforeAll) as if it were a new hook. The solution would be...well, not do that. [...]

More specifically it is a result of the static nature of these hooks, which is why we do not see this behaviour on the `beforeEach` and `afterEach` variants. This would have been a huge oversight on my part when initially implementing this feature in #282 , and could pose a potential performance issue for some folks, should they be doing any heavy work or network activity within.

### Solution

Funnily enough, it's a one-liner ... 😑 ... 🤣 . Nearly all changes are test related.

#### Source

- ensure that we re-initialize `static ?Closure $beforeAll` and `static ?Closure $afterAll` at constructor-time for each test case, within the `Testable` concern trait;

#### Tests

Due to how these hooks work (e.g., can be defined in multiple files, variable amount of tests using them, etc.), it's a little tricky to ensure they are only run once per file. The solution to this is outlined below.

- add a `calls` property to the `global $globalHook` object;
    - within here we keep a running total of hook calls, starting at 0;
- within each hook (we are testing), assert & increment it's call count;
    - call count is always incremented by 1, except in the tests themselves;
    - when asserting the call count, depending on the test we need to calculate and apply a delta/offset to our expected value;
        - see [`BeforeAllTest.php`](https://github.com/jordanbrauer/pest/blob/fix-static-hooks/tests/Hooks/BeforeAllTest.php#L6-L8) and [`AfterAllTest.php`](https://github.com/jordanbrauer/pest/blob/fix-static-hooks/tests/Hooks/AfterAllTest.php#L4-L8) header comments for full details

### CODE REVIEW

**Comment by jordanbrauer** [comment]
CC @nunomaduro @owenvoke @lukeraymonddowning 

Requesting a review for this bug fix when you find some time 🙂

File: src/Concerns/Testable.php [hunk 1]
```diff
@@ -73,6 +73,7 @@ public function __construct(Closure $test, string $description, array $data)
     {
         $this->__test        = $test;
         $this->__description = $description;
+        self::$beforeAll     = self::$afterAll = null;
```
Comment (lukeraymonddowning) [inline src/Concerns/Testable.php:76]: I know this is ripe coming from the guy who promotes higher order expectations inside higher order tests but should we split this over two lines for clarity?

**Comment by lukeraymonddowning** [comment]
@jordanbrauer just for my own understanding, if we add another test to the AfterEachHook test down the line, we'll need to update the BeforeEachHook test offset, yes?

File: src/Concerns/Testable.php (hunk 1 above)
Comment (jordanbrauer) [reply src/Concerns/Testable.php:76]: Probably, yeah. It is kind of easy to miss, if not careful. Two lines it is!

**Comment by jordanbrauer** [comment]
> if we add another test to the AfterEachHook test down the line, we'll need to update the BeforeEachHook test offset, yes?

@lukeraymonddowning , not quite, and thank goodness haha. The offset would only need to be updated if another test file was added to the `tests/Hooks/` directory.

For a better visual of why offset is `2`,

```
tests/Hooks
├── AfterAllTest.php           <-- no tests run before this one, so no offset
├── AfterEachTest.php          <-- 2 is our offset for the hook calls count
├── BeforeAllTest.php          <-- two tests run before this one
└── BeforeEachTest.php
```

This kind of sucks though still, because if we ever have a feature to randomize the test order (or if we currently do), it would fail for false positive. However, that being said, shouldn't be a problem since there are snapshot tests that indirectly care about the order.

Here I added some dummy tests that just expect true, but no changes to offset.

image

File: src/Concerns/Testable.php (hunk 1 above)
Comment (jordanbrauer) [reply src/Concerns/Testable.php]: Updated!

**Pull request merged**